docs/design/settings/extensions/modular-settings-extensions.md

```diff
@@ -0,0 +1,256 @@
+
+```bash
+$ schnauzer --required-extension 'hostname=v1' '{{ network.hostname }}'
+```
+
+#### Extending Handlebars Helpers
+Handlebars has a few utilities for applying simple processes to data in order to format it appropriately for your document; however, you often need to implement your own "helpers" for more complex text formatting.
+This is currently done in Bottlerocket by adding helper implementations to the `schnauzer` tool directly.
+Bottlerocket settings extensions will expose versioned handlebar helpers:
+
+```bash
+# hostname.toml setting extension config file
+[templating.v1] # template helpers are versioned
+helpers = [
+    "localhost_aliases"
+]
+
+
+$ ./hostname proto1 template-helper --version v1 --helper localhost_aliases 'arg1' 'arg2'
+#  rendered output returned as a string on stdout
+```
+
+Within the handlerbars template, the helper name will be prepended with `$EXTENSION_NAME.` to avoid collisions with other settings extensions.
+As an example, to use the `localhost_aliases` helper above from a template, your template may look something like this:
+
+```
+{{ hostname.localhost_aliases(arg1, arg2) }}
+```
+
+### Settings Generation
+Bottlerocket settings currently have the ability to compute a default value via something called a "settings generator".
+During boot, settings extensions are given a chance to populate settings with default values that are then persisted, effectively determining the setting value for the lifetime of that instance.
+
+Settings generation often depends on the presence of other settings.
+It's also possible for settings to have interdependencies on generation order, for example:
+
+* `foo.bar` depends on `baz.bot`
+* `boz.blop` depends on `foo.bar`
+
+In order to resolve this, settings generators are allowed to require that they be provided whatever data is currently applied to dependent settings.
+The generators are "re-entrant" in that they return a status dictating whether or not they are finished generating, or if they need more data to proceed.
+Bottlerocket will iteratively invoke generators until all generation has completed, or in cases where it seems likely that a deadlock has occurred.
+
+```toml
+# setting.toml requires the network and kubernetes settings before it can emit its own
+[generation.required]
+network = v1
+kubernetes = v2
+```
+
+## Settings Extension Development and the Settings SDK
+Settings extensions can be written in any programming language, so long as a symlink to a binary speaking the "extension protocol" lands in the right spot on the filesystem.
+That said, many settings extensions need to tackle the same challenges: How do we represent and implement settings versions? Migrations? Validation? Template utilities? Rather than forcing all extensions to implement this in isolation, we will create a shared library, called the "Settings SDK", which will be built in Bottlerocket's monorepo and published to crates.io.
```

Comment:
Let's not name this or say what repo it will be built from in this doc; it's not clear that we'd put this crate in the core repo vs. having another dedicated project for it.

I would like to avoid having a wide variety of different artifacts published from a single repo, since that complicates CI/CD workflows and the releases page.